### ISSUE TITLE
How many "positions" are there?

### ISSUE THREAD

**Issue by JJ** [Raku/doc#3774]
## Problem or new feature

From what I gather, a "position" is kinda like a context, except it refers to where the considered code is in relation with the rest of the code. However, I have only found it in the expression "term position". The fact that something can be, or not, in term position implies that there are at least two: term and *non-term*. But there might be others.

## Suggestions

For starters, clarify what is term position and index it correctly where found. Try and find out if there are other positions defined unambiguously.

**Comment by raiph** [comment]
**TL;DR** As I write this, my position is that there are likely at most two or three "positions" worth documenting. That is, "term position", "term position relative to "operator position", and "operator position" relative to "term position". Further, my position is that these are perhaps best left unspecified in prose, with links to [this doc section](https://docs.raku.org/language/operators#Operator_classification) serving as the definition. To the degree that's not considered a satisfactory solution, I suggest reading the rest of this comment.

----

I put *many* hours (iirc over 5 hours) into my research before drawing final tentative conclusions when I responded to [your question "What do you mean by "in term position"? in your 2019 SO comment](https://stackoverflow.com/questions/44101411/is-that-a-perl-6-hash-or-block/44102980#comment99215461_44102980).
and summarized my conclusions in comments that follow it (including a link to an attempt I made in an 2018 SO answer to define "term" and "position").

While I'd say my first comment, made a couple hours after your question, was especially weak, even including what seems to me as I write this to be a glaring mistake that I hereby retract ("which can't be further broken down into a list of values"), I'd say my research netted me enough to tentatively stand by the rest of what I wrote in those comments.

Fwiw I recall my researching including a good degree of contemplation following collation of my results after searching for relevant evidence in my mind, the web, design docs, IRC logs, RFCs, mailing lists from 2000s, perlmonks, perl5 docs, roast, and compiler source code.

I'll here highlight two elements of my comments:

* "I suggest asking Larry"

* A link to [an SO answer from 2018](https://stackoverflow.com/a/50779482/1077672) with the sections **What is a term?** and **How can one avoid the "Two terms in a row" error?**. Iirc I put some hours into research and contemplating how to write about such topics at that time too.

I think my comments and SO answer distilled enough that they're worth at least a quick read by anyone considering trying to resolve this issue.

> From what I gather, a "position" is kinda like a context, except it refers to where the considered code is in relation with the rest of the code.

That sounds about right to me.

But I advise *not* using the word *"context"* in any documentation of this. That's because "context" already has a deep and unrelated history in programming languages in general, in Larry Wall's languages in particular, and Raku. And while this unrelated notion is partly associated with strengths of Larry's approach, it's also strongly associated with what Larry and many others consider a major mistake. But most importantly, *none* of that usage of "context" is relevant to "term" or "position" in this, er, context.

> I have only found it in the expression "term position". The fact that something can be, or not, in term position implies that there are at least two: term and non-term. But there might be others.

----

As I write this, and let my intuition lead me to a conclusion that feels clear in this moment, including the awkwardness of my unflagged puns:

* The term "term" is likely best defined in terms of  the term "term position"; and

* The term "term position" is best defined as a syntactic position relative to "operator position"; and

* The terms "term position" and "operator position" are best defined as being relative to each other; and

* The table [in this doc section](https://docs.raku.org/language/operators#Operator_classification) is considered an admirably clear distillation of what needs to be said about all of these; and

* I suspect an attempt to turn that into prose will be fraught, especially given that I recall my research of related discussions from the 1990s yielded such incredibly vague discussion of these things that it's as if any attempt to precisely nail things down in prose is more confusing than not doing so.

**Comment by JJ** [comment]
"Term" and "operator" position make sense, since `*` will have different meanings depending on these. But I seem to vaguely remember also a distinction between "expression" and "statement". It's difficult to map these as "contexts", they're most likely understood as "positions". So that might be something else we might want to include in the prose. Which I think is a good thing, since it will avoid anyone else finding this in the future to expend as much time as you did...

**Comment by raiph** [comment]
> I seem to vaguely remember also a distinction between "expression" and "statement".

In a grammar like this (which I'm just making up as I write it):
```
grammar some-language {
    rule TOP { <statement>* %% ';' }
    rule statement { <expr> | <keyword> <block> }
    rule block { '{' <TOP> '}' }
}
```
`TOP`, `statement`, `expr`, `keyword`, and `block` are "positions".

The Raku grammar is waaaay more complicated than the above example. Afaict, as explained in my SO comments, what's explicitly referred to as a term (implicitly in a position) the word "term" in the error message "Two terms in a row" maps to the `termish` rule in the  main grammar.

> It's difficult to map these as "contexts"

I trust you read what I wrote above about "contexts".

(To recap, I think it would be a serious mistake to make *any* use of the specific word "context" in doc related to this issue, for the reasons I gave.)

> they're most likely understood as "positions".

Well, yes. I think that's why some linguists call them that ("grammatical slots" is an alternative), and Larry has used "positions" for more than 30 years.

But given that you opened this issue, perhaps they're most likely also *not* understood as "positions" by some folk!

Also, this might be why it seems Larry has resisted defining them, perhaps reflecting (his view that) they're better left as something implicitly understood than explicitly so.

> So that might be something else we might want to include in the prose.

I trust you read what I wrote above about use of prose.

(To recap, I question the wisdom of using prose in this instance, for the reasons I gave.)

> Which I think is a good thing, since it will avoid anyone else finding this in the future to expend as much time as you did.

I trust you read what I wrote above about not over-complicating things for the average reader.

(To recap, relying instead on linking to the existing doc table, because it's plausible, at least imo, that *prose* may make matters worse, for the reasons I gave.)

**Comment by raiph** [comment]
I forgot to mention that another source I recall checking was the Perl doc.

The "terms", "expressions", "statements", "positions" and so on are basically linguistic notions, which means they are relevant in general to natural and artificial languages such as programming languages. Nevertheless, their treatment and relevance in Perl and Raku are especially similar.

I recall checking Perl's doc, but don't recall finding anything useful, which is another of the factors that led me to think it's best left implicit other than the table.

But perhaps I'm wrong about that, and a quick check of Perl's doc by someone else right now seems to me like a smart move.

**Comment by pmichaud** [comment]
Raiph gives excellent responses to this thread -- I thought I'd add a few comments from a development/historical perspective that might help out a bit.

In development, we generally used "term position" and "operator position" as a shorthand for understanding what sorts of things could be consumed next by the expression parser.  However, things have evolved a fair bit since then, so that now I think the main alternate to "term position" would be "infix position".  (FWIW, I really like the "Operator classification table" in the docs -- I think that's really useful and this is the first time I've encountered it.)

For example, when parsing the beginning of an expression such as ```$x + 4```, the parse is considered to start in term position, because we're looking for a term.  In this case we'd consume the $x (a term, more properly term:sym<variable>), but other things can be terms as well, such as literals, function calls, circumfix expressions, and the like.

Once a term has been parsed, then we're in infix position, meaning infix-like things are expected next.  In this example, the operator would be infix:<+> .  Once the infix:<+> is parsed, we're back to "term position" because we're now expecting something that is term-like.

If we parse an expression up to an infix position and there's not a valid infix-ish token available next in the token stream, then the expression parse is considered complete and control returns to a "higher level" of parsing, such as a statement or signature.  (More on "statements" in a bit.)

Historically, we tended to use a phrase like "operator position" to refer to things that could be encountered after a term, which would include the postfix and postcircumfix operators as well as the infix operators.  But nowadays the postfix/postcircumfix/prefix/etc. operators are parsed directly as part of a term, so it's really just "infix position" that remains as the important one to distinguish.

As alluded to in the previous posts, the expression position is also how the parser recognizes the difference between things such as term:<*> (Whatever) and infix:<*> (multiplication), or prefix:<+> and infix:<+>.  

The main use of these from a user-perspective is in the error messages.  When the expression parser encounters a block in infix position, that often leads to the "two terms in a row" message.  Similarly, if the expression parser is in term position and encounters only infix operators, it can give a smarter error for that.  (These are the only usages I found in the Rakudo grammar.)

-----

With respect to "expression" versus "statement" -- yes, there's a distinction.  An expression is something like ```$x + 4``` or ```$z = $x * $y```; you can find them in many levels of parsing.

A statement is generally a "higher level construct", often involving keywords.  For example, an "if" statement is often considered something like:

   rule statement_if { 
      "if" <EXPR> <block> 
      [ "elsif" <EXPR> <block> ]* 
      [ "else" <block> ]?
   }

This defines an if statement as having a keyword "if", followed by an expression, followed by a block, followed by zero or more "elsif" clauses, followed by an optional else clause.  Other common statements are "unless", "for", "while", "loop", "use", etc., each of which has a specific syntax for the components of the statement.

Raku's if statement parsing is actually slightly more complex than the above example -- including pointy blocks and smarter error messages for things like "elif" and "else if" -- but not much more complex.

A standalone expression can itself be a statement.  In general, this means that if the ("higher-level") statement parser doesn't find a keyword initiating one of the known statements, then it looks for an expression to parse and if found uses that as a statement.

I don't know if "term position" and "infix position" need much heavy-duty explanation in the documentation for beginners -- certainly not as much as the above.  I think the operator classification table is really useful; there could potentially be an intermediate document somewhere that describes parsing and positions.

I hope this is useful somehow.

Pm

**Comment by JJ** [comment]
Quite useful. Thanks!
